recursion/program/src/challenger.rs

```diff
@@ -49,13 +49,31 @@ pub struct DuplexChallengerVariable<C: Config> {
 impl<C: Config> DuplexChallengerVariable<C> {
     /// Creates a new duplex challenger with the default state.
     pub fn new(builder: &mut Builder<C>) -> Self {
-        DuplexChallengerVariable::<C> {
+        let result = DuplexChallengerVariable::<C> {
             sponge_state: builder.dyn_array(PERMUTATION_WIDTH),
             nb_inputs: builder.eval(C::N::zero()),
             input_buffer: builder.dyn_array(PERMUTATION_WIDTH),
             nb_outputs: builder.eval(C::N::zero()),
             output_buffer: builder.dyn_array(PERMUTATION_WIDTH),
-        }
+        };
+
+        // Assert that the lengths of the arrays are equal to the permutation width.
+        builder.assert_usize_eq(PERMUTATION_WIDTH, result.sponge_state.len());
+        builder.assert_usize_eq(PERMUTATION_WIDTH, result.input_buffer.len());
+        builder.assert_usize_eq(PERMUTATION_WIDTH, result.output_buffer.len());
+
+        // Constrain the state of the challenger to contain all zeroes.
+        builder.assert_var_eq(result.nb_inputs, C::N::zero());
```

Comment:
Actually here it might be more efficient and correct to assign the value zero instead of asserting. Basically, here we are allocating an array but we did not initialize values. So you could do e.g.

builder.set(&result.sponge_state, I, C::F::zero())